### ISSUE TITLE
position of amp.init()

### ISSUE THREAD

**Issue by jinserk** [NVIDIA/apex#53]
Hi, thanks for sharing this great project!

I'd like to ask you about the position of `amp.init()`. Is it matter to call at the first of code or not? For example, can I `amp.init()` after all `nn.module` instances are generated?
Since there is no relation between the `amp_handle` and the `model`, I'm a little confused. In the case of `apex.fp16_utils`, it looks like I have to call `model = network_to_half(model)` before the training loop. But there is no procedure like that in the case of `amp`.

Thank you again!
Jin

**Comment by mcarilli** [comment]
Ideally, if you're using Amp, I recommend that you call `amp.init()` after you import the Torch namespace (and after any calls to `torch.cuda.set_device()`, if present), but before any module instances are generated.  You are correct that there's no explicit relationship between amp_handle and the model.  What amp does is patch functions in the torch namespace according to a whitelist-blacklist model.  For functions on the whitelist (safe/FP16-friendly), the patched function will downconvert incoming tensors to fp16.  For functions on the blacklist (potentially unsafe), the patched function will upconvert incoming tensors to fp32.

After `amp.init()`, any calls of `torch.*` functions executed by the model will route through these patched functions, and execute in fast fp16 when they can and safer fp32 when they must.  This is done transparently to the user.  Therefore, when using amp, you don't need to call `model.half()` or `network_to_half(model)`:  conversions happen internally, on the fly, when needed.

If you're using `fp16_utils` or `FP16_Optimizer`, you DO need to convert your model to half explicitly.  This is because `FP16_Optimizer` and `fp16_utils` do not alter the torch namespace at all:  Anything you explicitly declare to run in float will run in float, and anything you declare to run in half will run in half (aside from the gradient updates on the master weights, which are handled transparently).  In this way, `FP16_Optimizer` and `fp16_utils` are intended to give the user more control.

**Comment by jinserk** [comment]
Thank you @mcarilli for the kind answer. At first, I call `amp.init()` after the model instance generated, but had no error, then I thought the function wrapping after the model could make all submodules to fp32. However, I move the position of `amp.init()` to the very early of my python script, i.e. before the model generation, the running looks no difference at all. That's why I asked the question. I'm not sure whether `amp.init()` before any `model.forward()` calling is safe or not. As my understanding, it could be safe but you recommend to call `amp.init()` as earlier as possible. Am I correct?
Thank you again!

**Comment by mcarilli** [comment]
Again, amp operates by patching the torch namespace.  So if the module initialization does not call any `torch.*` functions/tensor ops, I expect that calling `amp.init()` before or after declaring the model should make no difference, as long as you call `amp.init()` at some point before the forward pass.

The reason I think it's safer to call `amp.init()` before initializing the model is because it's possible that in general, a model's constructor may call some `torch.*` functions, and you'd like those to be executed safely.
